segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,308 @@
+            try {
+                int appendLength = bufferViewToInt(readAppendOfLength(eventProcessor.segment, finalTruncationOffset, Integer.BYTES));
+                finalTruncationOffset += Integer.BYTES;
+                BufferView data = readAppendOfLength(eventProcessor.segment, finalTruncationOffset, appendLength);
+                readEvents.add(data);
+                finalTruncationOffset += data.getLength();
+            } catch (Exception e) {
+                // We got some data, do not throw, just stop looping.
+                if (!readEvents.isEmpty()) {
+                    break;
+                }
+                // TimeoutException is expected if there is no data in the Segment. Otherwise it may be a problem.
+                throw (e instanceof TimeoutException) ? new NoDataAvailableException() : new CompletionException(e);
+            }
+        }
+        return new EventsReadAndTruncationPoints(readEvents, currentOffset, finalTruncationOffset);
+    }
+
+    private CompletableFuture<Void> updateAndTruncateInternalSegment(EventProcessorImpl ep,
+                                                                     EventsReadAndTruncationPoints readResult,
+                                                                     Timer iterationTime) {
+        return updateLastProcessedEventOffset(ep.segment, readResult.getFinalTruncationOffset())
+                .thenCompose(v -> ep.segment.truncate(readResult.getFinalTruncationOffset(), OPERATION_TIMEOUT))
+                .thenAccept(v -> {
+                    // Decrement outstanding bytes and publish the last latency value after successful truncation.
+                    ep.outstandingBytes.addAndGet(-readResult.getReadBytes());
+                    ep.lastIterationLatency.set(iterationTime.getElapsedMillis());
+                });
+    }
+
+    private BufferView readAppendOfLength(DirectSegmentAccess segment, int offset, int length) throws Exception {
+        @Cleanup
+        ReadResult readResult = segment.read(offset, length, OPERATION_TIMEOUT);
+        ReadResultEntry entry = readResult.next();
+        entry.requestContent(OPERATION_TIMEOUT);
+        return entry.getContent().get(OPERATION_TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    private int bufferViewToInt(BufferView appendLengthContent) throws IOException {
+        return ByteBuffer.wrap(appendLengthContent.getReader().readAllBytes()).getInt();
+    }
+
+    private ByteArraySegment intToByteArraySegment(int intValue) {
+        return new ByteArraySegment(ByteBuffer.allocate(Integer.BYTES).putInt(intValue).array());
+    }
+
+    private CompletableFuture<Integer> getLastProcessedEventOffset(DirectSegmentAccess segment) {
+        return segment.getAttributes(Collections.singletonList(LAST_PROCESSED_EVENT_OFFSET), true, OPERATION_TIMEOUT)
+                      .thenApply(attributes -> attributes.getOrDefault(LAST_PROCESSED_EVENT_OFFSET, 0L).intValue());
```

Comment:
Why `intValue`? What happens after your segment exceeds 2GB?

Reply:
This has been removed, now we do not use a special attribute anymore.